pymc3/distributions/multivariate.py

```diff
@@ -473,28 +473,31 @@ class Multinomial(Discrete):
     Parameters
     ----------
     n : int or array
-        Number of trials (n > 0).
+        Number of trials (n > 0). If n is an array its shape must be (N,) with 
+        N = p.shape[0]
     p : one- or two-dimensional array
         Probability of each one of the different outcomes. Elements must
-        be non-negative and sum to 1 along the last axis. They will be automatically
-        rescaled otherwise.
+        be non-negative and sum to 1 along the last axis. They will be 
+        automatically rescaled otherwise.
     """
 
     def __init__(self, n, p, *args, **kwargs):
         super(Multinomial, self).__init__(*args, **kwargs)
 
         p = p / tt.sum(p, axis=-1, keepdims=True)
+        n = np.squeeze(n) # works also if n is a tensor
 
-        lst = range(self.shape[-1])
         if len(self.shape) > 1:
             m = self.shape[-2]
             try:
                 assert n.shape == (m,)
-            except AttributeError:
-                n *= tt.ones(m)
+            except (AttributeError, AssertionError):
+                n = n * tt.ones(m)
+            self.n = tt.shape_padright(n)
+            self.p = p if p.ndim > 1 else tt.shape_padleft(p)
+        elif n.ndim == 1:
```

Comment:
it looks a little funny that the conditions here are `if len(self.shape) > 1: ... elif n.ndim == 1: ... else: ...`  is there a reason that both conditions can't be assertions about `self.shape`, or both about `n.ndim`?

Reply:
Yes, `self.shape` doesnt work when there is `Observed`. I can't find a way to overwrite the default, so instead I am checking `n.dim` to interprete the shape.

I think we should refactor it to have a stricter shape (like what you said on twitter, `n` and `p` being both 2D and Observed being 2D or 3D). Something like:
```
n -> (k, 1)
p -> (m, k)
observerd -> (m, k) or (n, m, k)
```

Reply:
I was poking around, and there doesn't seem to be a theano `atleast_2d`.  The best solution would be for `numpy` or `scipy` to support such vectorized sampling...